### PR TITLE
Mergeable heuristic in KCFGSemantics for merging node

### PR DESCRIPTION
**Pull request by Stevengre**
This is a simple PR for merging nodes, offering a semantics-level heuristic to partition nodes that can be merged.

### CODE REVIEW

**Comment by ehildenb** [comment]
I don't think it makes sense to merge this without also having some tests of it somehow. Do you have an example of IMP node merging worknig, or another simple definition?

**Comment by Stevengre** [comment]
I tested this on an actual Solidity program a few days ago. However, this PR primarily introduces an interface to define the heuristic, which can't be tested without the merge_node function. Since the new interface results in many test changes, I plan to submit a separate PR for those.

File: pyk/src/pyk/kcfg/semantics.py [hunk 1]
```diff
@@ -34,6 +34,11 @@ def custom_step(self, c: CTerm) -> KCFGExtendResult | None: ...
 
     """Implement a custom semantic step."""
 
+    @abstractmethod
+    def is_mergeable(self, c1: CTerm, c2: CTerm) -> bool: ...
```
Comment (Stevengre) [inline pyk/src/pyk/kcfg/semantics.py:38]: This PR only introduces an interface. Its semantics will be determined by the merge_node function, which will be provided in another PR.

**Comment by ehildenb** [comment]
Well we can wait for the merging funcitonality to be ready to go, and merged, then merge this new interface and some tests of it with the node merging. This will trigger needed changes downstream, because each of the downstream semantics will have to now implement this interface, and we don't know if it's fully working yet.

Once https://github.com/runtimeverification/k/pull/4617 and https://github.com/runtimeverification/k/pull/4621 are merged, we can then have this PR introduce both the `merge_nodes` predicate, the main algorithm for merging KCFG nodes, and some tests over one of the semantics.

**Comment by Stevengre** [comment]
Thank you for your suggestion! I'll follow your advice.

**Comment by ehildenb** [comment]
Subsumed by: https://github.com/runtimeverification/k/pull/4628
